File: Spiral Matrix.cpp
class Solution {
public:
    vector<int> spiralOrder(vector<vector<int>>& matrix) {
        vector<int> answer;
        int a = matrix.size();
        int b = matrix[0].size();
        int total = a*b;
        int rt= b-1, rb= a-1, lb=0, lt=1;
        int i=0, j=0;
        while(total>0){
            while(j<=rt && total>0){
                answer.push_back(matrix[i][j]);
                j++;
                total--;
            }i++;
            j--;
            if(total==0)
                break;
            while(i<=rb && total>0){
                answer.push_back(matrix[i][j]);
                i++;
                total--;
            }j--;
            i--;
            if(total==0)
                break;
            while(j>=lb && total>0){
                answer.push_back(matrix[i][j]);
                j--;
                total--;
            }i--;
            j++;
            if(total==0)
                break;
            while(i>=lt && total>0){
                answer.push_back(matrix[i][j]);
                i--;
                total--;
            }j++;
            i++;
            rt--;
            rb--;
            lb++;
            lt++;
        }
        return answer;
    }
};

File: Spiral Matrix.py
class Solution:
    def spiralOrder(self, matrix: List[List[int]]) -> List[int]:
        spiral = []
        row = len(matrix)
        col = len(matrix[0])
        total = row*col
        rt, rb, lt, lb = col-1, row-1, 1, 0
        i, j = 0, 0
        
        while total:
            while j<=rt and total>0:
                spiral.append(matrix[i][j])
                j += 1
                total -= 1
            i += 1
            j -= 1
            if total == 0:
                break
            
            while i<=rb and total>0:
                spiral.append(matrix[i][j])
                i += 1
                total -= 1
            i -= 1
            j -= 1 
            if total == 0:
                break

            while j>=lb and total>0:
                spiral.append(matrix[i][j])
                j -= 1
                total -= 1
            j += 1
            i -= 1
            if total == 0:
                break

            while i >= lt and total>0:
                spiral.append(matrix[i][j])
                i -= 1
                total -= 1
            i += 1
            j += 1
            rt -= 1
            rb -= 1
            lt += 1
            lb += 1
            
        return spiral
